state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -1246,154 +1220,248 @@ pub mod pallet {
-								swap.output_asset()
-							);
-						}
-					},
-					SwapRequestState::IngressEgressFee => {
-						Self::deposit_event(Event::<T>::SwapRequestCompleted { swap_request_id });
+					};
 
-						if swap.output_asset() ==
-							ForeignChain::from(swap.output_asset()).gas_asset()
-						{
-							T::IngressEgressFeeHandler::accrue_withheld_fee(
-								swap.output_asset(),
-								output_amount,
+					if is_gas_swap {
+						*gas_swap_state = GasSwapState::OutputReady { gas_budget: output_amount };
+					} else {
+						// The executed swap must be for the principal amount:
+						dca_state.accumulated_output_amount += output_amount;
+
+						// Schedule any remaining DCA chunks:
+						dca_state.scheduled_chunk_swap_id = if dca_state.remaining_chunks > 0 {
+							let chunk_input_amount = dca_state.prepare_next_chunk();
+
+							let swap_id = Self::schedule_swap(
+								request.input_asset,
+								request.output_asset,
+								chunk_input_amount,
+								swap.swap.refund_params.clone(),
+								SwapType::CcmPrincipal,
+								swap_request_id,
 							);
+
+							Some(swap_id)
 						} else {
-							log_or_panic!(
-								"IngressEgressFee swap should not be to non-gas asset: {:?}",
-								swap.output_asset()
+							None
+						};
+
+						// See if we still need to schedule the gas swap
+						if let GasSwapState::ToBeScheduled { gas_budget, other_gas_asset } =
+							gas_swap_state
+						{
+							*gas_swap_state = Self::schedule_ccm_gas_swap(
+								swap_request_id,
+								request.input_asset,
+								*other_gas_asset,
+								*gas_budget,
 							);
 						}
-					},
-				}
+					}
+
+					// Process outcome if there aren't any more scheduled swaps,
+					// otherwise put the request state back:
+					if let (GasSwapState::OutputReady { gas_budget }, None) =
+						(gas_swap_state, dca_state.scheduled_chunk_swap_id)
+					{
+						Self::egress_for_swap(
+							swap_request_id,
+							dca_state.accumulated_output_amount,
+							request.output_asset,
+							output_address.clone(),
+							Some((ccm_deposit_metadata.clone(), *gas_budget)),
+							false, /* refund */
+						);
+
+						true
+					} else {
+						false
+					}
+				},
+				SwapRequestState::Regular { output_address, dca_state } => {
+					if dca_state.remaining_chunks > 0 {
+						let chunk_input_amount = dca_state.prepare_next_chunk();
+						dca_state.accumulated_output_amount += output_amount;
+
+						Self::schedule_swap(
+							request.input_asset,
+							request.output_asset,
+							chunk_input_amount,
+							// TODO: recompute swap params depending on the output amount we got
+							swap.swap.refund_params.clone(),
+							SwapType::Swap,
+							request.id,
+						);
+
+						false
+					} else {
+						debug_assert!(dca_state.remaining_input_amount == 0);
+
+						Self::egress_for_swap(
+							swap_request_id,
+							output_amount + dca_state.accumulated_output_amount,
```

Comment:
Ok I see - it's because of this? 

Seems more consistent to add it above and then egress the `accumulated_output_amount`, no?

Reply:
See the answer above. It is a trade-off. We have to choose between consistency and rewriting if/esle with two separate ifs on the same condition, which seems ugly too.